### PR TITLE
btest/cluster/websocket: Add cert-less test

### PR DESCRIPTION
**Pull request by awelzel**
@rsmmr mentioned on #3852, that ixwebsocket should support encryption without certs.

Indeed, it does work on my system with the following configuration:
```
	# Enable encryption, but don't require authentication (and not even certificates).
	#
	# This is the old Broker::listen_websocket() default behavior.
	local tls_options = Cluster::WebSocketTLSOptions(
		$cert_file="",
		$key_file="",
		$ca_file="NONE",
		$ciphers="AECDH-AES256-SHA@SECLEVEL=0:AECDH-AES256-SHA:P-384",
	);

	local ws_server_options = Cluster::WebSocketServerOptions(
		$listen_addr=127.0.0.1,
		$listen_port=to_port(getenv("WEBSOCKET_PORT")),
		$tls_options=tls_options,
	);

	Cluster::listen_websocket(ws_server_options);
```

Lets test if this works on all platforms, so we can point users that do actually want this at the test for an example.

### CODE REVIEW

**Comment by awelzel** [comment]
Not digging into the `ubuntu22_spicy` issue, but @bbannier / @0xxon maybe you have a hunch what it might be?

```
[#1] core.pppoe-over-qinq ... failed on retry #2
  % 'btest-diff conn.log' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  #separator \x09
  #set_separator	,
  #empty_field	(empty)
  #unset_field	-
  #path	conn
  #open	2025-06-05-13-18-11
  #fields	ts	uid	id.orig_h	id.orig_p	id.resp_h	id.resp_p	proto	service	duration	orig_bytes	resp_bytes	conn_state	local_orig	local_resp	missed_bytes	history	orig_pkts	orig_ip_bytes	resp_pkts	resp_ip_bytes	tunnel_parents	ip_proto
  #types	time	string	addr	port	addr	port	enum	string	interval	count	count	string	bool	bool	count	string	count	count	count	count	set[string]	count
  1523351398.449222	CHhAvVGS1DHFjwGM9	1.1.1.1	20394	2.2.2.2	443	tcp	ssl	273.626833	11352	4984	SF	F	F	0	ShADdtaTTtFf	44	25283	42	13001	-	6
  #close	2025-06-05-13-18-11
  == Diff ===============================
  --- /dev/fd/63	2025-06-05 13:18:11.917440912 +0000
  +++ /dev/fd/62	2025-06-05 13:18:11.918440910 +0000
  @@ -6,5 +6,5 @@
   #open XXXX-XX-XX-XX-XX-XX
   #fields	ts	uid	id.orig_h	id.orig_p	id.resp_h	id.resp_p	proto	service	duration	orig_bytes	resp_bytes	conn_state	local_orig	local_resp	missed_bytes	history	orig_pkts	orig_ip_bytes	resp_pkts	resp_ip_bytes	tunnel_parents	ip_proto
   #types	time	string	addr	port	addr	port	enum	string	interval	count	count	string	bool	bool	count	string	count	count	count	count	set[string]	count
  -XXXXXXXXXX.XXXXXX	CHhAvVGS1DHFjwGM9	1.1.1.1	20394	2.2.2.2	443	tcp	-	273.626833	11352	4984	SF	F	F	0	ShADdtaTTtFf	44	25283	42	13001	-	6
  +XXXXXXXXXX.XXXXXX	CHhAvVGS1DHFjwGM9	1.1.1.1	20394	2.2.2.2	443	tcp	ssl	273.626833	11352	4984	SF	F	F	0	ShADdtaTTtFf	44	25283	42	13001	-	6
   #close XXXX-XX-XX-XX-XX-XX
  =======================================
```

**Comment by 0xxon** [comment]
ooooh - I bet that has something to do with the dpd change. I totally forgot to run the spicy ssl test on that - which is, I guess, why we have that ci.

**Comment by 0xxon** [comment]
It totally is caused by #4362. It's one of the corner cases, where there is an ssl connection that has an error that happens relatively late in the connection.

I still have to dig down what exactly is causing this change of behaviour. I will try to be quick - but I'm not sure it will be today.

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	Cluster::subscribe("/zeek/event/my_topic");
+
+	# Enable encryption, but don't require authentication (and not even certificates).
+	#
+	# This is the old Broker::listen_websocket() default behavior.
+	local tls_options = Cluster::WebSocketTLSOptions(
+		$cert_file="",
+		$key_file="",
+		$ca_file="NONE",
+		$ciphers="AECDH-AES256-SHA@SECLEVEL=0:AECDH-AES256-SHA:P-384",
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek:45]: Nit:

```suggestion
		$ciphers=join_string_vec(vector("AECDH-AES256-SHA@SECLEVEL=0", "AECDH-AES256-SHA", "P-384"), ":"),
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+# @TEST-PORT: XPUB_PORT
+# @TEST-PORT: XSUB_PORT
+# @TEST-PORT: LOG_PULL_PORT
+# @TEST-PORT: WEBSOCKET_PORT
+#
+# @TEST-EXEC: cp $FILES/zeromq/cluster-layout-simple.zeek cluster-layout.zeek
+# @TEST-EXEC: cp $FILES/zeromq/test-bootstrap.zeek zeromq-test-bootstrap.zeek
+#
+# @TEST-EXEC: zeek -b --parse-only manager.zeek
+# @TEST-EXEC: python3 -m py_compile client.py
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek:17]: The test is driven by the client, no need to put it into the background.

```suggestion
# @TEST-EXEC: python3 client.py >out
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+event Cluster::websocket_client_added(info: Cluster::EndpointInfo, subscriptions: string_vec)
+	{
+	print "Cluster::websocket_client_added", subscriptions;
+	}
+
+event Cluster::websocket_client_lost(info: Cluster::EndpointInfo, code: count, reason: string)
+	{
+	print "Cluster::websocket_client_lost";
+	terminate();
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek:73]: If we did a clean `exit(0)` here we would expect nothing on stderr and wouldn't need to baseline it.

```suggestion
	exit(0);
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	}
+# @TEST-END-FILE
+
+
+# @TEST-START-FILE client.py
+import asyncio, json, os, socket, ssl, time
+from websockets.asyncio.client import connect
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek]: Nit:

```suggestion
import asyncio
import json
import os
import socket
import ssl
import time

from websockets.asyncio.client import connect
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        "type": "data-message",
+        "topic": topic,
+        "@data-type": "vector",
+        "data": [
+            {"@data-type": "count", "data": 1},  # Format
+            {"@data-type": "count", "data": 1},  # Type
+            {"@data-type": "vector", "data": [
+                { "@data-type": "string", "data": "ping"},  # Event name
+                { "@data-type": "vector", "data": [  # event args
+                    {"@data-type": "string", "data": f"python-websocket-client"},
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek]:
```suggestion
                    {"@data-type": "string", "data": "python-websocket-client"},
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                    {"@data-type": "count", "data": c},
+                ], },
+            ], },
+        ],
+    }
+
+async def run():
+    ctx = ssl.create_default_context()
+    ctx.set_ciphers("AECDH-AES256-SHA@SECLEVEL=0:AECDH-AES256-SHA:P-384")
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek:115]: Nit:

```suggestion
    ctx.set_ciphers(':'.join(["AECDH-AES256-SHA@SECLEVEL=0", "AECDH-AES256-SHA", "P-384"]))
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            ev = pong["data"][2]["data"]
+            print("topic", pong["topic"], "event name", ev[0]["data"], "args", ev[1]["data"])
+
+def main():
+    for _ in range(100):
+        try:
+            asyncio.run(run())
+            break
+        except ConnectionRefusedError:
+            time.sleep(0.1)
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek]: Let's capture errors to connect, e.g.,

```suggestion
    for _ in range(100):
        try:
            asyncio.run(run())
            return
        except ConnectionRefusedError:
            time.sleep(0.1)

    raise RuntimeError("could not connect to Zeek endpoint")
```

File: testing/btest/cluster/websocket/tls-nocert.zeek [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+# @TEST-EXEC: python3 -m py_compile client.py
+#
+# @TEST-EXEC: btest-bg-run manager "ZEEKPATH=$ZEEKPATH:.. && CLUSTER_NODE=manager zeek -b ../manager.zeek >out"
+# @TEST-EXEC: btest-bg-run client "python3 ../client.py >out"
+#
+# @TEST-EXEC: btest-bg-wait 30
+# @TEST-EXEC: btest-diff ./manager/out
+# @TEST-EXEC: btest-diff ./manager/.stderr
+# @TEST-EXEC: btest-diff ./client/out
+# @TEST-EXEC: btest-diff ./client/.stderr
```
Comment (bbannier) [inline testing/btest/cluster/websocket/tls-nocert.zeek]: Suggest to not baseline these at all

- `manager/.stderr` should have nothing relevant
- `client/out` only has debug information, if the client finishes running we are fine
- `client/.stderr` is automatically emitted on stderr if we do not put the client in the background

**Review by bbannier — APPROVED** [review]
LGTM, only left a couple of nits.

File: testing/btest/cluster/websocket/tls-nocert.zeek (hunk 1 above)
Comment (awelzel) [reply testing/btest/cluster/websocket/tls-nocert.zeek:45]: Will skip this one.

File: testing/btest/cluster/websocket/tls-nocert.zeek (hunk 2 above)
Comment (awelzel) [reply testing/btest/cluster/websocket/tls-nocert.zeek:17]: Yeah, good point.

File: testing/btest/cluster/websocket/tls-nocert.zeek (hunk 8 above)
Comment (awelzel) [reply testing/btest/cluster/websocket/tls-nocert.zeek]: Fair point and I like it!

File: testing/btest/cluster/websocket/tls-nocert.zeek (hunk 3 above)
Comment (awelzel) [reply testing/btest/cluster/websocket/tls-nocert.zeek:73]: I'll stick with the terminate as it may flush some state, but no need for the baseline, agree.
